### ISSUE TITLE
surface_hausdorff_distance does not work for completely filled image

### ISSUE THREAD

**Issue by FeynmanDNA** [InsightSoftwareConsortium/SimpleITK-Notebooks#453]
Hi, I found that the `surface_hausdorff_distance()` in https://github.com/InsightSoftwareConsortium/SimpleITK-Notebooks/blob/master/Python/34_Segmentation_Evaluation.ipynb fails to work when the image is completely filled up.

This looks to be because `SignedMaurerDistanceMap` fails to compute the distance map when the image is completely filled (e.g. all 1s).

For example, if you print the distance map and surface for a filled up toy image of all 1s:

**Fails when image is completely filled:**

```python
image1 = sitk.Image([2, 2], sitk.sitkUInt8)
image1[:, :] = 1

image1:
[[1 1]
 [1 1]]
```

Will give you:
```py
segmented_distance_map = 
[[1.8446743e+19 1.8446743e+19]
 [1.8446743e+19 1.8446743e+19]]
segmented_surface =
 [[0 0]
 [0 0]]
```

**But it works if the image is padded:**

```python
# pad the image1:
[[0 0 0 0]
 [0 1 1 0]
 [0 1 1 0]
 [0 0 0 0]]
```

```py
segmented_distance_map = 
[[1.4142135 1.        1.        1.4142135]
 [1.        0.        0.        1.       ]
 [1.        0.        0.        1.       ]
 [1.4142135 1.        1.        1.4142135]]
segmented_surface = 
[[0 0 0 0]
 [0 1 1 0]
 [0 1 1 0]
 [0 0 0 0]]
```

I found that as long as the image is completely filled, `surface_hausdorff_distance()` does not work. Once it is not a completely filled up image (with padding or just pad one axis), it will work again. So maybe can try to pad the ref and pred images just in case?

**Comment by zivy** [comment]
Hello @FeynmanDNA,

Thanks for the bug report. 

**TL;DR**: resolved by [this PR](https://github.com/InsightSoftwareConsortium/SimpleITK-Notebooks/pull/455) 

**Long explanation**: This has to do with the surface computation, so issue occurs when the segmentation touches the image border. The case where the segmentation fills the whole image is a specific instance of the problem.

Adding the following cell at the end of the notebook illustrates the problem:
```
ref = sitk.Image([32,32], sitk.sitkUInt8)
seg = sitk.Image([32,32], sitk.sitkUInt8)

ref[0:10,0:10] = 1
seg[1:10, 1:10] = 1

sitk.Show(ref*255, "ref") # multiply by 255 so immediately visible in Fiji and no need to play with intensity for viewing 
sitk.Show(seg*255, "seg")

refpad = sitk.ConstantPad(ref,[1]*ref.GetDimension(), [1]*ref.GetDimension())
segpad = sitk.ConstantPad(seg,[1]*ref.GetDimension(), [1]*ref.GetDimension())

print(str(surface_hausdorff_distance(ref, seg))
)
print(str(surface_hausdorff_distance(refpad, segpad))
)
```

**Comment by FeynmanDNA** [comment]
Hi @zivy Thank you very much for the info. Glad to have helped!

**Comment by FeynmanDNA** [comment]
@zivy Hmm, I found that strangely, the seemingly harmless padding of the image actually caused the surface distance calculation to be different.

Before padding, i.e., the still public notebook results: https://insightsoftwareconsortium.github.io/SimpleITK-Notebooks/Python_html/34_Segmentation_Evaluation.html#Segmentation-Representation-and-the-Hausdorff-Distance

gives:
```py
HausdorffDistanceImageFilter result (reference1-segmentation): 7.0
Surface Hausdorff result (reference1-segmentation): 6.0
HausdorffDistanceImageFilter result (reference2-segmentation): 3.605551275463989
Surface Hausdorff result (reference2-segmentation): 8.9442720413208
```

**But with the padding PR/fix, the surface hausdorff result is different for both ref1 and ref2 (6.3 vs 6) and (9.4 vs 8.9):**

```py
HausdorffDistanceImageFilter result (reference1-segmentation): 7.0
Surface Hausdorff result (reference1-segmentation): 6.324555397033691
HausdorffDistanceImageFilter result (reference2-segmentation): 3.605551275463989
Surface Hausdorff result (reference2-segmentation): 9.433980941772461
```

I found that increasing the padding size has no effect, **but strangely, a `[0] *dim` fake padding also gives a different result from before the fix**:

```py
    dim = reference_segmentation.GetDimension()
    reference_segmentation = sitk.ConstantPad(
        reference_segmentation, [0] * dim, [0] * dim
    )
    seg = sitk.ConstantPad(seg, [0] * dim, [0] * dim)
```

```py
HausdorffDistanceImageFilter result (reference1-segmentation): 7.0
Surface Hausdorff result (reference1-segmentation): 6.324555397033691
HausdorffDistanceImageFilter result (reference2-segmentation): 3.605551275463989
Surface Hausdorff result (reference2-segmentation): 9.433980941772461
```

It looks like something happened with `sitk.ConstantPad` that caused the surface distance calculation to change. Nonetheless, it is strange that the distances returned are different from before the fix.

Shall I open a new issue for this? 

Thanks!

**Comment by FeynmanDNA** [comment]
@zivy After comparing the before and after code, I found that the fix might have a bug calculating the `distance_map`. **Previously it used the segmentation mask to calculate the distance map, but now it (wrongly) uses the surface contour to calculate the distance map (`reference_segmentation` instead of `reference_surface`)** :

```py
    reference_distance_map = sitk.Abs(
        sitk.SignedMaurerDistanceMap(
            reference_surface, squaredDistance=False, useImageSpacing=True
        )
    )
```

It should be:
```py
    reference_distance_map = sitk.Abs(
        sitk.SignedMaurerDistanceMap(
            reference_segmentation, squaredDistance=False, useImageSpacing=True
        )
    )
```

There is nothing wrong with `ConstantPad`. The difference in calculation is probably due to the bug in wrongly using `reference_surface` and `segmented_surface`.

**Comment by zivy** [comment]
Hello @FeynmanDNA,

The rendered notebook you are looking at is very old, older than [this bugfix](https://github.com/InsightSoftwareConsortium/SimpleITK-Notebooks/commit/4a3967e5edeb6f746e4c79d53b416a2489ba8346) (I need to update all the renderings). If you run the current notebook and just comment out the couple of lines that deal with padding you will see that you get the same result with and without padding. The results are different from the rendered notebook because that still had the old bug.
